core-rust/p2p/src/rocks_db.rs

```diff
@@ -0,0 +1,245 @@
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+use std::path::PathBuf;
+
+use node_common::rocksdb::{
+    ColumnFamilyDescriptor, DB, Options,
+};
+
+use crate::engine_prelude::*;
+use crate::address_book_components::AddressBookNodeId;
+use crate::column_families::{AddressBookCf, HighPriorityPeersCf, MigrationStatusCf, SafetyStoreCf};
+use crate::traits::node::{AddressBookStore, HighPriorityPeersStore, MigrationStore, SafetyStateStore};
+use crate::migration::{MigrationId, MigrationStatus};
+use node_common::store::rocks_db::*;
+use node_common::store::typed_cf_api::*;
+
+/// A listing of all column family names used by the Node.
+///
+/// This is directly needed to initialize the column families within the DB, but is also a nice
+/// place to link to all of them (please see the documentation of each CF to learn about its
+/// business purpose and DB schema) and to put the important general notes regarding all of them
+/// (see below).
+///
+/// **Note on the key encoding used throughout all column families:**
+/// We often rely on the RocksDB's unsurprising ability to efficiently list entries sorted
+/// lexicographically by key. For this reason, our byte-level encoding of certain keys (e.g.
+/// [`StateVersion`]) needs to reflect the business-level ordering of the represented concept (i.e.
+/// since state versions grow, the "last" state version must have a lexicographically greatest key,
+/// which means that we need to use a constant-length big-endian integer encoding).
+///
+/// **Note on the name strings:**
+/// The `NAME` constants defined by `*Cf` structs (and referenced below) are used as database column
+/// family names. Any change would effectively mean a ledger wipe. For this reason, we choose to
+/// define them manually (rather than using the `Into<String>`, which is refactor-sensitive).
+
+const ALL_ADDRESS_BOOK_COLUMN_FAMILIES: [&str; 2] = [
+    AddressBookCf::DEFAULT_NAME,
+    HighPriorityPeersCf::DEFAULT_NAME,
+];
+const ALL_SAFETY_STORE_COLUMN_FAMILIES: [&str; 1] = [
+    SafetyStoreCf::DEFAULT_NAME,
+];
+const ALL_MIGRATION_STORE_COLUMN_FAMILIES: [&str; 1] = [
+    MigrationStatusCf::DEFAULT_NAME,
+];
+
+pub type ActualAddressBookDatabase = AddressBookDatabase<DirectRocks>;
+pub type ActualSafetyStoreDatabase = SafetyStoreDatabase<DirectRocks>;
+pub type ActualMigrationDatabase = MigrationDatabase<DirectRocks>;
+
+/// A RocksDB-backed persistence layer for address book.
+pub struct AddressBookDatabase<R> {
+    /// Underlying RocksDB instance.
+    rocks: R,
+}
+
+/// A RocksDB-backed persistence layer for safety store.
+pub struct SafetyStoreDatabase<R> {
```

Comment:
I'd expect the `SafetyStore` to be in a separate consensus crate. But I'm happy for us to move it at a later date if we just want to get this ticket in.

Reply:
I think further refactoring (including moving p2p JNI part) should be done at the next stage, after discussion and making decision on final structure.